### PR TITLE
Create attribute within the new product MVP

### PR DESCRIPTION
**Pull request by louwie17**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

This allows users to create a new attribute if the name they are searching for doesn't exist within the attribute list.
This is currently dependent on #34999 to be merged, which is why it's pointing at that branch.

https://user-images.githubusercontent.com/2240960/195870763-a609958f-b3ab-4515-a498-25660423b5c9.mp4

Address the creation part in #34331  .

### How to test the changes in this Pull Request:

1. Load this branch, build it, and make sure you have the `new-product-management-experience` feature flag enabled (you can do so using the latest version of the Beta tester within the mono repo).
2. Go to **Products > Add new (MVP)** and fill out some of the initial details like name and price
3. Scroll down and click **Add first attribute** in the attributes section
4. Now write a random name in the attribute search field in the table, of an attribute that doesn't exist.
5. The 'Create "your new attribute name"' should show as an option
6. Click that and the new item should show as selected and the term field should focus
7. With `localStorage.setItem( 'debug', 'wc-admin:*' );` run prior, it should show that a `product_attribute_add_custom_attribute` has been fired.
8. Start adding attribute terms, this dropdown should initial be blank and any text you put in will show up as a `Create "text"` option, allowing you to add several new items.
9. Click **Add** the custom attribute should show correctly in the main list, now hit `Update` or `Publish`
10. Go to old product editor and load the same product you just saved, check the attributes tab and make sure the custom one's show up correctly.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> run changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce-admin/client/products/fields/attribute-input-field/attribute-input-field.tsx
Comment (louwie17) [inline plugins/woocommerce-admin/client/products/fields/attribute-input-field/attribute-input-field.tsx]: This is just indented as I had to add a Fragment around this in order to add the create modal

File: plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx]: Could this be a partial of `HydratedAttributeType` or does it need to have a separate `{ id: undefined }`?

File: plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx]: It would be great to have this as a util method so we could also test it.

E.g., `parseAttribute( val: string | Attribute )`

File: plugins/woocommerce-admin/client/products/fields/attribute-input-field/attribute-input-field.tsx
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/fields/attribute-input-field/attribute-input-field.tsx]: To reduce the number of `attribute.id === -99` checks, maybe we could create a util method `isNewAttribute( attribute )` that handles this check for us.

File: plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx]: I know that `formAttr` was not added in this PR, but this confused me for a minute as I thought it was a "form attribute."   Would `attribute` be a more accurate name?  Or is there something special about these that is related to the form?

**Review by joshuatf — DISMISSED** [review]
Very nice work, Lourens!  Testing well on my end; left a few minor comments, but nothing blocking and pre-approving this PR.

Not a result of this PR, but I noticed a couple of oddities around styling:

1. The modal appears off when editing an existing attribute
Screen Shot 2022-11-16 at 7 44 44 AM

2.  The loading indicator is slightly disorienting and jumpy, making it challenging to click an item while loading occurs.  This loading indicator might be better as the select control's icon instead of in the list (cc @jarekmorawski).  Though this may already be planned and blocked by the select control icon issue.
Screen Shot 2022-11-16 at 7 43 45 AM

**Comment by jarekmorawski** [comment]
> The loading indicator is slightly disorienting and jumpy, making it challenging to click an item while loading occurs. This loading indicator might be better as the select control's icon instead of in the list (cc @jarekmorawski). Though this may already be planned and blocked by the select control icon issue.

Definitely. The latest approach is to replace the search icon with a loading indicator, but we'll work on that as part of another issue.

File: plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx
Comment (louwie17) [reply plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx]: I wanted to have a clear differentiator between when a attribute was set and when it wasn't. Using a Partial of HydratedAttributeType would make that a bit more difficult and we did have to manually check if all the attributes exist on that property to deem it as the HydratedAttributeType in several places.

Comment (louwie17) [reply plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx]: I could potentially just try `null`, let me see if that works well.

Comment (louwie17) [reply plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx]: I updated it to using `HydratedAttributeType | null` so that things are more clear.

File: plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx
Comment (louwie17) [reply plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx]: Good suggestion, I did do this as part of the https://github.com/woocommerce/woocommerce/pull/35100/commits/fd8af0d7fe3a812dbff3f5ace8077e099f6298f8 commit (it includes the other changes as well)

File: plugins/woocommerce-admin/client/products/fields/attribute-input-field/attribute-input-field.tsx
Comment (louwie17) [reply plugins/woocommerce-admin/client/products/fields/attribute-input-field/attribute-input-field.tsx]: Done as part of: https://github.com/woocommerce/woocommerce/pull/35100/commits/2dcbcd50c61e66e3a28be45f78d9fc2101674947

**Comment by louwie17** [comment]
@joshuatf this should be ready for re-review, I addressed all your changes.
Also in relation to the two issues you found:

> The modal appears off when editing an existing attribute

This should be fixed now after the rebase as I fixed this as part of the create attribute PR -> https://github.com/woocommerce/woocommerce/pull/35132 (at-least I can't reproduce it anymore)

> The loading indicator is slightly disorienting and jumpy, making it challenging to click an item while loading occurs. This loading indicator might be better as the select control's icon instead of in the list (cc @jarekmorawski). Though this may already be planned and blocked by the select control icon issue.

As @jarekmorawski mentioned this is blocked by the other issue in flight to allow us to move the loading indicator within the input field.

File: plugins/woocommerce-admin/client/products/fields/attribute-input-field/attribute-input-field.tsx
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/fields/attribute-input-field/attribute-input-field.tsx]: This could also use the new `isNewAttributeListItem` right?

File: plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx:101]: I think ideally we could use a ref here instead of dom selectors, but we can save that for a follow-up issue as I know that wasn't really a part of this PR.

File: plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx]: Awesome!

**Review by joshuatf — DISMISSED** [review]
Thanks for the changes, @louwie17!  This is still testing well for me and changes look great.

Left one more comment about using refs instead of dom selectors for a possible follow-up issue.  Otherwise, LGTM!  🚢

File: plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx
Comment (louwie17) [reply plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx:101]: Yeah currently the SelectControl component doesn't allow us to pass in a `ref`, I could definitely add that though as a separate PR

File: plugins/woocommerce-admin/client/products/fields/attribute-input-field/attribute-input-field.tsx
Comment (louwie17) [reply plugins/woocommerce-admin/client/products/fields/attribute-input-field/attribute-input-field.tsx]: Nice catch, not sure how I missed that, fixed here: https://github.com/woocommerce/woocommerce/pull/35100/commits/9a7f07b479e490434f630bb78a541959d40f7874

**Comment by louwie17** [comment]
@joshuatf could I get a re-review/approval I just updated the one last spot that needed to use the helper function.

**Review by joshuatf — APPROVED** [review]
Changes look good!  LGTM :shipit:

**Pull request merged**
